### PR TITLE
Report number of physically running processes to status.csv

### PR DESCRIPTION
**Pull request by almiliMSFT**
The new column name is `RunningProcesses`.  This is similar to `ExternalProcesses`, except that `ExternalProcesses` counts the number of processes executing under `PipExecutor`, whereas `RunningProcesses` counts the number of actual user-specified processes physically running (any child processes spawned by those main pip processes are not included here).

### CODE REVIEW

File: Public/Src/Engine/Scheduler/PipExecutor.cs [hunk 1]
```diff
@@ -1134,7 +1134,7 @@ private static void MakeSharedOpaqueOutputIfNeeded(IPipExecutionEnvironment envi
         /// <param name="state">the pip scoped execution state</param>
         /// <param name="pip">The pip to execute</param>
         /// <param name="fingerprint">The pip fingerprint</param>
-        /// <param name="processIdListener">Callback to call when the process is actually started</param>
+        /// <param name="processIdListener">Callback to call when the process is actually started (PID is passed to it) and when the process exit (negative PID is passed to it)</param>
```
Comment (almiliMSFT) [inline Public/Src/Engine/Scheduler/PipExecutor.cs]:
```suggestion
        /// <param name="processIdListener">Callback to call when the process is actually started (PID is passed to it) and when the process exited (negative PID is passed to it)</param>
```

File: Public/Src/Engine/UnitTests/Processes/SandboxedProcessTest.cs [hunk 2]
```diff
@@ -1085,6 +1085,31 @@ public async Task TempAccessesAreUnexpectedByDefault()
                 string.Join(Environment.NewLine, result.AllUnexpectedFileAccesses.Select(p => p.GetPath(Context.PathTable))));
         }
 
+        [Fact]
+        public async Task ProcessStartExitIsReportedToListener()
+        {
+            var srcFile = CreateSourceFile();
+
+            var fam = new FileAccessManifest(Context.PathTable);
+            fam.ReportFileAccesses = true; // We explicitly do not set ReportAccess (testing catchall reporting)
```
Comment (almiliMSFT) [inline Public/Src/Engine/UnitTests/Processes/SandboxedProcessTest.cs]:
```suggestion
            fam.ReportFileAccesses = false;
```

File: Public/Src/Tools/Execution.Analyzer/Analyzers.Core/Xldb/XldbProtobufExtensions.cs [hunk 3]
```diff
@@ -7,6 +7,7 @@
 using BuildXL.Scheduler.Tracing;
 using BuildXL.Utilities;
 using BuildXL.Xldb.Proto;
+using Microsoft.VisualBasic;
```
Comment (almiliMSFT) [inline Public/Src/Tools/Execution.Analyzer/Analyzers.Core/Xldb/XldbProtobufExtensions.cs]:
```suggestion
```

File: Public/Src/Engine/Scheduler/Tracing/ExecutionLog.Events.cs [hunk 4]
```diff
@@ -1280,6 +1285,7 @@ public void Serialize(BinaryLogger.EventWriter writer)
             writer.Write(LookupRunning);
 
             writer.Write(ExternalProcesses);
```
Comment (semihokur) [inline Public/Src/Engine/Scheduler/Tracing/ExecutionLog.Events.cs]: We might rename ExternalProcesses to PipExecutorProcesses. your call?

File: Public/Src/Engine/Scheduler/Distribution/LocalWorker.cs [hunk 5]
```diff
@@ -21,6 +23,13 @@ public sealed class LocalWorker : Worker
         /// </summary>
         public ConcurrentDictionary<PipId, Unit> CurrentlyExecutingPips = new ConcurrentDictionary<PipId, Unit>();
 
+        /// <summary>
+        /// The number of pips that are currently running (i.e., the associated pip process is still alive and running)
+        /// </summary>
+        public int CurrentlyRunningPipCount => m_currentlyRunningPipCount;
```
Comment (semihokur) [inline Public/Src/Engine/Scheduler/Distribution/LocalWorker.cs]: can we rename it to RunningProcesses to be more consistent with the column name?

**Comment by narasamdya** [comment]
Why do we need `RunningProcesses` counter? Is it not the same (or almost the same) as the number of executing process pip?

File: Public/Src/Engine/Processes/SandboxedProcess.cs [hunk 6]
```diff
@@ -492,6 +492,8 @@ private async Task OnProcessExitingAsync()
 
         private async Task OnProcessExited()
         {
+            m_processIdListener?.Invoke(-ProcessId);
```
Comment (narasamdya) [inline Public/Src/Engine/Processes/SandboxedProcess.cs]: >m_processIdListener [](start = 12, length = 19)

Does it have to go as low as `SandboxedProcess`? Can it be tracked in `SandboxedProcessPipExecutor`?

File: Public/Src/Engine/Scheduler/Distribution/LocalWorker.cs (hunk 5 above)
Comment (narasamdya) [inline Public/Src/Engine/Scheduler/Distribution/LocalWorker.cs]: >m_currentlyRunningPipCount [](start = 47, length = 26)

I think we need `Volatile.Read` for reading currently running pip count.

**Comment by almiliMSFT** [comment]
> Why do we need `RunningProcesses` counter? Is it not the same (or almost the same) as the number of executing process pip?

That's why: to be able to see if it is indeed close to number of executing process pips (because it seems that in some cases it's not)

File: Public/Src/Engine/Scheduler/Distribution/LocalWorker.cs (hunk 5 above)
Comment (almiliMSFT) [reply Public/Src/Engine/Scheduler/Distribution/LocalWorker.cs]: will change

File: Public/Src/Engine/Processes/SandboxedProcess.cs (hunk 6 above)
Comment (almiliMSFT) [reply Public/Src/Engine/Processes/SandboxedProcess.cs]: Good point.  It's more accurate if it's here, but I don't think it needs to be that accurate, so I'll move it.

File: Public/Src/Engine/Scheduler/Distribution/LocalWorker.cs (hunk 5 above)
Comment (almiliMSFT) [reply Public/Src/Engine/Scheduler/Distribution/LocalWorker.cs]: will do

File: Public/Src/Engine/Scheduler/Tracing/ExecutionLog.Events.cs (hunk 4 above)
Comment (almiliMSFT) [reply Public/Src/Engine/Scheduler/Tracing/ExecutionLog.Events.cs]: sounds good

File: Public/Src/Engine/Scheduler/Tracing/ExecutionLog.Events.cs [hunk 7]
```diff
@@ -1215,7 +1215,12 @@ public struct StatusEventData : IExecutionLogEventData<StatusEventData>
         /// <summary>
         /// Number of externally running processes
```
Comment (almiliMSFT) [inline Public/Src/Engine/Scheduler/Tracing/ExecutionLog.Events.cs]:
```suggestion
        /// Number of processes running under PipExecutor
```

File: Public/Src/Engine/Processes/SandboxedProcess.cs (hunk 6 above)
Comment (almiliMSFT) [reply Public/Src/Engine/Processes/SandboxedProcess.cs]: done.

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 8]
```diff
@@ -736,8 +740,6 @@ private bool SandboxedProcessNeedsExecuteExternal
 
         private bool ShouldSandboxedProcessExecuteExternal
             => SandboxedProcessNeedsExecuteExternal
-               // Process does not talk to BuildXL server.
-               && m_processIdListener == null
```
Comment (narasamdya) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: >&& m_processIdListener == null [](start = 15, length = 30)

I like this!

**Pull request merged**